feat: update substrait to `0.57.1`

wackywendell opened the pull request:
The validator seemed to be a bit out of date with the main substrait repo, so I decided to see what it would take to update it to current substrait. And... it wasn't too bad.

For new features, I simply added them to the validator as `NotYetImplemented`, so they should emit warnings (not errors).

The only place I added actual support was for `IntervalDayToSeconds`, which should now accept the `precision`/`subseconds` variety, and not just the deprecated `microseconds`.

EpsilonPrime commented on rs/src/parse/expressions/literals.rs:
```diff
@@ -1049,6 +1091,26 @@ fn parse_null(x: &substrait::Type, y: &mut context::Context) -> diagnostic::Resu
 
+fn parse_value(
+    x: &substrait::expression::literal::user_defined::Val,
+    y: &mut context::Context,
```
ctx?

wackywendell replied:
Yes, can do. The rest of this file uses `x` and `y`, but `ctx` is better 👍

EpsilonPrime commented on rs/src/parse/expressions/literals.rs:
```diff
@@ -38,8 +38,13 @@ enum LiteralValue {
     Binary(Vec<u8>),
 
-    /// May be used only for IntervalYearToMonth and IntervalDayToSecond.
-    Interval(i64, i64),
```
Should we reject use of Interval as deprecated at this point or just return an error as we would likely do once the old fields are removed?

wackywendell replied:
I'm still new to this code base, but I think this is an internal class that is not exported; it does not have a 1:1 match with the public protobuf files. So there isn't a way to do `Interval` (and not `IntervalYearToMonth` and `IntervalDayToSecond`) in protobufs right now - there's nothing to deprecate or return an error on.

Does that make sense? (Or am I missing something?)

EpsilonPrime commented on rs/src/parse/relations/set.rs:
```diff
@@ -68,6 +68,19 @@ pub fn parse_set_rel(x: &substrait::SetRel, y: &mut context::Context) -> diagnos
+                y,
+                Warning,
+                NotYetImplemented,
+                "Variant {:?} not yet supported",
```
should probably include word set in the message

wackywendell replied:
Yes, done 👍

EpsilonPrime reviewed (commented):
Thanks for working on getting this moving into the right direction.  This is a great first step towards bringing the validator up to where it should be.

I suspect we will want to avoid releases until we address all of the basic requirements of the latest version (such as enforcing that function names have their accepted argument types in their names) but that shouldn't hold up this first, great step.

mbrobbel commented on rs/src/parse/expressions/literals.rs:
```diff
@@ -677,32 +677,74 @@ fn parse_interval_year_to_month(
-    proto_primitive_field!(x, y, microseconds);
+    proto_primitive_field!(x, ctx, seconds);
+    proto_primitive_field!(x, ctx, subseconds);
+    proto_field!(x, ctx, precision_mode);
+
+    let (subseconds, prec) = match x.precision_mode {
+        // TODO: Is the default microseconds?
```
I would assume `precision_mode` is a required field.

mbrobbel commented on rs/src/parse/relations/join.rs:
```diff
@@ -96,12 +112,19 @@ pub fn parse_join_rel(x: &substrait::JoinRel, y: &mut context::Context) -> diagn
-        (JoinType::Anti, false) => "Anti",
-        (JoinType::Single, true) => "Filtered single",
-        (JoinType::Single, false) => "Single",
+        (JoinType::LeftSemi, true) => "Filtered semi",
+        (JoinType::LeftSemi, false) => "Semi",
+        (JoinType::LeftAnti, true) => "Filtered anti",
+        (JoinType::LeftAnti, false) => "Anti",
+        (JoinType::LeftSingle, true) => "Filtered single",
+        (JoinType::LeftSingle, false) => "Single",
```
```suggestion
        (JoinType::LeftSemi, true) => "Filtered left semi",
        (JoinType::LeftSemi, false) => "Left semi",
        (JoinType::LeftAnti, true) => "Filtered left anti",
        (JoinType::LeftAnti, false) => "Left anti",
        (JoinType::LeftSingle, true) => "Filtered left single",
        (JoinType::LeftSingle, false) => "Left single",
```

mbrobbel commented on rs/src/parse/relations/join.rs:
```diff
@@ -146,22 +169,44 @@ pub fn parse_join_rel(x: &substrait::JoinRel, y: &mut context::Context) -> diagn
+                                    or throw an error."
+                        .to_string()
+                }
+                JoinType::LeftMark => "TODO: What is LeftMark?".to_string(),
+                JoinType::RightMark => "TODO: What is RightMark?".to_string(),
```
https://substrait.io/relations/logical_relations/#join-types

wackywendell replied:
Ah, thanks!

mbrobbel commented on rs/src/parse/relations/mod.rs:
```diff
@@ -47,7 +47,27 @@ fn parse_rel_type(x: &substrait::rel::RelType, y: &mut context::Context) -> diag
+                x.proto_oneof_variant()
+            );
+            Ok(())
+        } //
+          // _ => Ok(()),
```
```suggestion
        }
```

wackywendell replied:
This is not a new comment, but one moved around before; it relates to this:
https://github.com/substrait-io/substrait-validator/blob/e325a89f0bfc294b29dcd010c17f3bdef0cba727/rs/src/parse/relations/mod.rs#L33-L36

I'd prefer it on its own line, but `rustfmt` doesn't like that.

I can clarify a bit at this point in the code? Or alternatively, I guess I could just drop it - the comment above is clear enough?

mbrobbel replied:
Ah missed that. Let's just keep it like this.

mbrobbel reviewed (commented):
Thanks for working on this!

wackywendell commented:
Thanks for the comments! I'm working through them now 👍

wackywendell commented:
OK - I've addressed all the comments from code review. There's a couple above that I'm not sure I fully handled - let me know what you think!

And thank you both for your quick reviews, @EpsilonPrime and @mbrobbel !

mbrobbel commented on rs/src/parse/expressions/literals.rs:
```diff
@@ -677,32 +677,76 @@ fn parse_interval_year_to_month(
-    proto_primitive_field!(x, y, microseconds);
+    proto_primitive_field!(x, ctx, seconds);
+    proto_primitive_field!(x, ctx, subseconds);
+    proto_field!(x, ctx, precision_mode);
```
I think you can use:
```suggestion
    proto_required_field!(x, ctx, precision_mode);
```

wackywendell replied:
So I didn't require this because I'm not sure if we should require precision mode. In particular, requiring precision mode when no subseconds (or even seconds) are specified seems unnecessary; in fact, there's [a TPCH test](https://github.com/substrait-io/substrait-validator/blob/8b22bdbebccc27e38b0dafe73d5dae3f54fc78e1/tests/tests/tpc-h/tpc-h01.yaml#L265-L266) that uses `intervalDayToSecond: { days: 120 }` which has no precision mode set, but seems pretty reasonable since neither seconds nor subseconds are set.

Does that make sense? Does that seem reasonable?

wackywendell commented:
Thanks for the reviews! The last thing seems to be the commit lints. I can do some squashing and rebasing to address, although I'm generally hesitant to change history on publicly available commits; it will make the review comments above hard to track. Is that worth doing here?

mbrobbel commented:
> Thanks for the reviews! The last thing seems to be the commit lints. I can do some squashing and rebasing to address, although I'm generally hesitant to change history on publicly available commits; it will make the review comments above hard to track. Is that worth doing here?

No, it doesn't matter. This repo uses squash merge, so it'll use the PR title as merge commit message.
